### PR TITLE
Install TURN/STUN server with coturn on Debian/Ubuntu

### PR DESCRIPTION
**Pull request by n-se**
As there are not enough default examples for IP addresses as mentioned in #34, I have used here
```
Client public IPv4: 78.47.166.55 (hetzner.com IPv4)
Client private IPv4: 192.168.0.10
```
in lines 28, 29 and 184-189

I have not found a complete tutorial online, this one is mixed from several tutorials and also contains own experiences. The installation and configuration routine is fully tested.

### CODE REVIEW

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
+# Install TURN/STUN Server on Debian/Ubuntu with coturn
+
+## Introduction
+
+We will install a basic TURN/STUN server with [coturn](https://github.com/coturn/coturn).
+The TURN Server is a VoIP media traffic NAT traversal server and gateway. It can be used as a general-purpose network traffic TURN server and gateway, too.
```
Comment (frommMoritz) [inline tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]: What's meant by this? Explain it a little bit less tetchier

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+_turn._udp.example.com.   14400 IN SRV  5  50 3478 turn1.example.com.
+_turn._udp.example.com.   14400 IN SRV  10 50 3478 turn2.example.com.
+_turn._tcp.example.com.   14400 IN SRV  5  50 3478 turn1.example.com.
+_turn._tcp.example.com.   14400 IN SRV  10 50 3478 turn2.example.com.
+_turns._tcp.example.com.  14400 IN SRV  5  50 5349 turn1.example.com.
+_turns._tcp.example.com.  14400 IN SRV  10 50 5349 turn2.example.com.
+```
+Do not forget to setup DNS records (A and AAAA) for `turn1.example.com` and turn2.example.com` as mentioned above.
+
+Theoretically you could also deploy a second A and AAAA records for `turn.example.com` and `stun.example.com` but this is not the recommended way.
```
Comment (frommMoritz) [inline tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]:
```suggestion
Theoretically you could also deploy a second A and AAAA record for `turn.example.com` and `stun.example.com` but this is not the recommended way.
```

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+## Step 2 - Installation
+
+First we will need to install `coturn` on our server.
+`apt-get install coturn`
+
+For the moment we will stop `coturn`
+`systemctl stop coturn`
+
+## Step 3 - Configuring
```
Comment (frommMoritz) [inline tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]: This part needs to be resorted. You should first name the configuration file, then explain what to change in it.

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+Open the file and remove the `#` in front of `TURNSERVER_ENABLED=1` in order to enable the TURN server.
```
Comment (frommMoritz) [inline tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]:
```suggestion
In order to enable the TURN server, open the file and remove the `#` in front of `TURNSERVER_ENABLED=1`.
```

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+The configuration file is located at `/etc/turnserver.conf`. First we move the original file with
```
Comment (frommMoritz) [inline tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]:
```suggestion
The configuration file is located at `/etc/turnserver.conf`. First we move the original version of this file using
```

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+```bash
+mv /etc/turnserver.conf /etc/turnserver.conf.orig
+```
+
+than we open`/etc/turnserver.conf` with an editor and paste the following configuration file:
```
Comment (frommMoritz) [inline tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]:
```suggestion
than we open/create `/etc/turnserver.conf` in an editor of our choice and paste the following configuration:
```

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+#log-file=/dev/null
+
+no-loopback-peers
+no-multicast-peers
+
+proc-user=turnserver
+proc-group=turnserver
+```
+
+Create the secret with
```
Comment (frommMoritz) [inline tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]: What's this secret need for?

```suggestion
Generate the secret by using
```

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+```bash
+sed -i "s/replace-this-secret/$(openssl rand -hex 32)/" /etc/turnserver.conf
+```
+
+## Step 4 - Register Let's Encrypt certificates
```
Comment (frommMoritz) [inline tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]:
```suggestion
## Step 4 - Create Let's Encrypt certificates
```

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+```bash
+chmod a+x ./certbot-auto
+./certbot-auto certonly --standalone --rsa-key-size 4096 -m holu@example.com -d turn.example.com
+```
+You will get an email 30 (and 7) days before the certificates expire to holu@example.com.
+
+## Step 5 - Start and test the TURN/STUN server
+
+To start coturn execute
```
Comment (frommMoritz) [inline tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]:
```suggestion
We will start coturn by executing
```

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+`systemctl start coturn`
+and now test with `systemctl status coturn` if coturn is running.
```
Comment (frommMoritz) [inline tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]:
```suggestion
using `systemctl status coturn` we can test if coturn is running.
```

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+Remove the default server from the list.
+The field for “STUN or TURN URI” should look like this:
+
+`stun:stun.example.com`
+
+but you can also use the IP:
+
+`stun:10.0.0.1:3478`
+
+You can also use port `5349` to test with SSL/TLS encryption here.
```
Comment (frommMoritz) [inline tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]:
```suggestion
To test SSL/TLS encryption, use port `5349`.
```

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+
+The output should look similar to this:
+```
+Time    Component       Type    Foundation     Protocol Address         Port        Priority
+0.019	1	        host             0          UDP	192.168.0.10	41904	126 | 32512 | 255
+0.024	1	        host             2          TCP	192.168.0.10	9	125 | 32704 | 255
+0.027	2	        host             0          UDP	192.168.0.10	53438	126 | 32512 | 254
+0.030	2	        host             2          TCP	192.168.0.10	9	125 | 32704 | 254
+0.134	1	        srflx            1          UDP	78.47.166.55	41904	100 | 32543 | 255
+0.152	2	        srflx            1          UDP	78.47.166.55	53438	100 | 32543 | 254
```
Comment (frommMoritz) [inline tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]: What's up with those IPs?

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+0.153	Done
+```
+
+If you get a timeout with `Not reachable?` then probably a firewall is blocking the connection. Check again if the ports, `3478` and `5349`, for the TURN server are opened and if you can reach it externally via telnet or nmap.
```
Comment (frommMoritz) [inline tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]:
```suggestion
If you get a timeout with `Not reachable?` the connection is probably blocked by a firewall. Check again if the ports `3478` and `5349` for the TURN server are open. Also test if you can reach it from your computer or another server via telnet or nmap.
```

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+
+If everything works as excepted replace `log-file=/var/log/turn.log` with `log-file=/dev/null` to disable IP address logging for best privacy.
```
Comment (frommMoritz) [inline tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]:
```suggestion
If everything works as excepted, replace `log-file=/var/log/turn.log` with `log-file=/dev/null` in your configuration file. This disables IP address logging and is best for the privacy of your  users.
```

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md (hunk 1 above)
Comment (n-se) [reply tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]: It is already explained in a very basic way. People who do not understand these two sentences probably do not need a TURN or STUN server.

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md (hunk 7 above)
Comment (n-se) [reply tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]: You do not really create the certificate, you have to register it. You only say `create` if you self sign it.

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md (hunk 12 above)
Comment (n-se) [reply tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]: It is your clients internal (private) IP and your public IP address. This is STUNs job. Should be clear if you wanna deploy a STUN server.

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md (hunk 8 above)
Comment (n-se) [reply tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]: You do not really create the certificate, you have to register it. You only say `create` if you self sign it.

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md (hunk 14 above)
Comment (n-se) [reply tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]: I have rewritten this a bit in 949290c1db06a703fb9426a5dab969128bd48269

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md (hunk 3 above)
Comment (n-se) [reply tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]: Fixed with 949290c1db06a703fb9426a5dab969128bd48269

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md (hunk 1 above)
Comment (fdrechsler) [reply tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]: Yes, you should be more precise about that.
Maybe add a Mermaid Diagram, they are supported in the Community Tutorial Renderer.

Something like that: 

```mermaid
sequenceDiagram
    participant ClientA
    participant TURN Server
    participant ClientB
    opt p2p not possible
        TURN Server->>+ClientA: Forward Signals
        TURN Server->>-ClientB: Forward Signals
    end
    ClientA->ClientB: p2p communication
    Note over ClientA,ClientB: p2p/e2e SRTP Encrypted Streaming ( TURN Server not involved )
```

Will render to 
image

Comment (fdrechsler) [reply tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]: Yes, you should be more precise about that.
Maybe add a Mermaid Diagram, they are supported in the Community Tutorial Renderer.

Something like that: 

```mermaid
sequenceDiagram
    participant ClientA
    participant TURN Server
    participant ClientB
    opt p2p not possible
        TURN Server->>+ClientA: Forward Signals
        TURN Server->>-ClientB: Forward Signals
    end
    ClientA->ClientB: p2p communication
    Note over ClientA,ClientB: p2p/e2e SRTP Encrypted Streaming ( TURN Server not involved )
```

Will render to 
image

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+- [Jitsi Meet](https://jitsi.org/jitsi-meet/)
+- [Matrix Server](https://matrix.org/)
+- [ejabberd](https://www.ejabberd.im/)
+
+
+If you self-host one of these applications you can, for example, use your own TURN/STUN server there.
+
+**Prerequisites**
+
+- At least a  small virtual server, e.g. [CX11](https://www.hetzner.com/cloud), with a root shell.
```
[the remark this answers is not in the document]
Comment (n-se) [reply tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]: Thanks, I have tried `#cx11` but that didn't worked.

File: tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md (hunk 1 above)
Comment (fdrechsler) [reply tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]: > probably do not need a TURN or STUN server
People don't know what they don't know :) Somebody 100% certain about the Topic would probably not read the Tutorial.

Comment (n-se) [reply tutorials/install-turn-stun-server-on-debian-ubuntu-with-coturn/01.en.md]: Diagram added as suggested in 5d77c0c2764fc7a6f69a122572e26b85a8dd0fdd

**Pull request merged**